### PR TITLE
refactor(engine): addEventListener restriction options param

### PR DESCRIPTION
**Pull request by aputinski**
## Details

Closes https://github.com/salesforce/lwc/issues/1824

## Does this PR introduce breaking changes?

* ✅ `No, it does not introduce breaking changes.`

## GUS work item
W-7446526

### CODE REVIEW

File: packages/@lwc/engine/src/framework/restrictions.ts [hunk 1]
```diff
@@ -456,6 +423,37 @@ function getLightningElementPrototypeRestrictionsDescriptors(
                 return originalIsConnectedGetter.call(this);
             },
         }),
+        addEventListener: generateDataDescriptor({
+            value(
+                this: HTMLElement,
```
Comment (caridy) [inline packages/@lwc/engine/src/framework/restrictions.ts:428]: this is not HTMLElement here, it is `LightningElement`

File: packages/@lwc/engine/src/framework/restrictions.ts [hunk 2]
```diff
@@ -456,6 +423,37 @@ function getLightningElementPrototypeRestrictionsDescriptors(
...
+                assert.invariant(
+                    !isInvokingRender,
+                    `${vmBeingRendered}.render() method has side effects on the state of ${componentTag} by adding an event listener for "${type}".`
+                );
+                assert.invariant(
+                    !isUpdatingTemplate,
+                    `Updating the template of ${vmBeingRendered} has side effects on the state of ${componentTag} by adding an event listener for "${type}".`
+                );
+                // TODO [#420]: this is triggered when the component author attempts to add a listener
+                // programmatically into a lighting element node
```
Comment (ravijayaramappa) [inline packages/@lwc/engine/src/framework/restrictions.ts:445]:
```suggestion
                // programmatically into a lighting component instance
```

File: packages/@lwc/engine/src/framework/restrictions.ts (hunk 1 above)
Comment (caridy) [reply packages/@lwc/engine/src/framework/restrictions.ts:428]:
```suggestion
                this: LightningElement,
```

**Comment by caridy** [comment]
Ok, this code is now ready, but after discussing the details with @ravijayaramappa, we looked at the code, and in fact, the custom element is being patched, eliminating the ability to use options when using synthetic shadow. We have found a couple of ways to allow this, but we will have to put some time to solve this.

**Comment by pmdartus** [comment]
Closing this PR based on @caridy's feedback.

**Pull request closed without merging**